### ISSUE TITLE
Alarm not ringing or ringing later than specified

### ISSUE THREAD

**Issue by Rafee-M** [vicolo-dev/chrono#183]
**Describe the bug**
The alarm seems to ring ~10 minutes late or doesn't ring at all sometimes

**To Reproduce**
1. Set an alarm more than 30 minutes away
2. Wait

**Expected behavior**
Ring at specified time

**Smartphone Information**
 - Device: OnePlus 9RT
 - OS: Android 13
 - App version: 4.1 (clean install)

**Additional context**
I had enabled foreground, background usage, turned off optimization, enabled auto launch. Still the alarm ringing is a hit or miss. In contast, I gave [Simple Clock](https://github.com/SimpleMobileTools/Simple-Clock) only foreground usage and it still seems to ring at specified times

**Screenshots**

**Example 1:** At 8:30 AM an alarm was set for 9:35 AM. It rang at 9:39 only after unlocking the phone and opening Chrono

![Untssitled](https://github.com/vicolo-dev/chrono/assets/69535896/aa819d88-0a54-449b-b68a-973a453c05f2)
![Untitled](https://github.com/vicolo-dev/chrono/assets/69535896/12fb1924-6992-43cd-8b56-3afe0b8f54ce)

**Example 2:** A notification for upcoming alarm at 11:15 AM (set at 10:30 AM) was receieved. But the alarm failed to ring. Opening the app also didn't make the alarm ring as well.
![Screenshot_2024-04-19-11-29-13-07](https://github.com/vicolo-dev/chrono/assets/69535896/f2b4e367-0e15-438f-8182-beeafadd9455)

https://github.com/vicolo-dev/chrono/assets/69535896/43aac825-14e0-4826-ae0e-75e262609b9e

Battery settings for Chrono:

![Untiasatled](https://github.com/vicolo-dev/chrono/assets/69535896/df89046b-4c85-4983-9584-082ffdb81a55)

**Comment by AhsanSarwar45** [comment]
Have you tried everything listed here: https://dontkillmyapp.com/oneplus

And here:
https://oneplus.gadgethacks.com/how-to/disable-setting-if-notifications-are-delayed-your-oneplus-0192639/

Especially the Advanced Optimization thing. 

I don't own a OnePlus device but can you also make sure that 'Show on Lock Screen' or something similar and 'Show Fullscreen Notifications' or something similar is enabled for the app?

**Comment by Rafee-M** [comment]
App Ops and Uses permissions of Simple Clock are given below. The only thing stood out to me was the "Start Foreground" option. And Chrono seems to have quite a bit more Uses permissions.

**App Ops:**

<img src="https://github.com/vicolo-dev/chrono/assets/69535896/14fafa5d-5232-4819-8663-8efb275edb16" width="180" height="400">

**Uses Permissions:**
1. <img src="https://github.com/vicolo-dev/chrono/assets/69535896/3264c5af-978a-4921-a268-5599d5782468" width="182" height="350">
2. <img src="https://github.com/vicolo-dev/chrono/assets/69535896/3e9f9ffd-ae7e-49c0-a09d-a5811338426a" width="181" height="350">

**Comment by AhsanSarwar45** [comment]
Thanks, I'll take a look what SimpleClock does differently. Also, when did this issue first started for you? After 0.4.0/0.4.1 update or before that?

**Comment by Rafee-M** [comment]
> Also, when did this issue first started for you? After 0.4.0/0.4.1 update or before that?

I had some reliability issues as well, so that is why I was going back and forth between Chrono and Simple Clock.

Also, I have resolved the issue with the alarm by disabling the "Battery Optimization" option inside [App Manager](https://f-droid.org/packages/io.github.muntashirakon.AppManager/). Not sure what setting it enables/disables, but it does use root permissions. 
For now, the issue seems to have been resolved for me. Feel free to close it! Also, there seems to be an issue with the timer as well. I'll try to do some more testing and create a new issue.

The [App Manager](https://f-droid.org/packages/io.github.muntashirakon.AppManager/) option:

https://github.com/vicolo-dev/chrono/assets/69535896/92d6aefc-d709-43de-a3ac-c56d8b7cb4f5

**Comment by jfmcbrayer** [comment]
I have this same behavior on a Moto G7 running LineageOS, I'm afraid. It only started happening after upgrading from LineageOS 20 (Android 13) to LineageOS 21 (Android 14). Battery optimization is off, and notifications are on.

What specifically happens is that the alarm does not go off at all, but the upcoming alarm notification is shown. The alarm does not go off after opening the app.

I don't know if this will help - Android 14 seems to have broken a lot of things that use background processing for me, from alarms to offline music players.

**Comment by AhsanSarwar45** [comment]
> I don't know if this will help - Android 14 seems to have broken a lot of things that use background processing for me, from alarms to offline music players.

Don't have a real Android 14 device at the moment but I've tested it on a stock android 14 emulator, and it works well there. So probably a real device specific optimization.

**Comment by aronovgj** [comment]
I had this happen on my Pixel 7 pro. Did not happen before 0.4.0. 
I disabled battery optimization in app manager, will see if that helps.

**Comment by aronovgj** [comment]
The alarm didn't go off 3 nights in a row. It works after not touching the phone for about half an hour. Tinkering with app manager did not help. Trying to figure out if the issue is connected with turning on rising volume.

**Comment by AhsanSarwar45** [comment]
I mentioned it in another issue, but will also mention it here. I am currently looking at adding a foreground service (optional in the settings), which will show a peristent notification and hopefully prevent the app from being killed. It will help in fixing alarm reliability issues. 

Also, if possible, can you guys run the benchmark in DontKillMyApp from the app store and report the results? 

It would also be helpful if you can you guys can suggest opensource apps that work correctly in the background on your devices, and whether they do it through a persistent notification (foreground service).

**Comment by jfmcbrayer** [comment]
I think the apps I have that are *not* having problems with being killed on Android 14 all have foreground services (and thus persistent notifications). I'm talking about music apps here, though, not clock apps. The only other clock app I have tried is LineageOS stock, which rings, but only silently.

**Comment by aronovgj** [comment]
Don't kill my app returns 100%.

K9 Mail runs well on my phone. I think it also uses a foreground service.

I deleted the old alarm that was failing and created a new one without rising volume and alarm worked this morning. Might be luck. Not sure.

**Comment by jfmcbrayer** [comment]
On my device, DontKillMyApp returns 100%.

I also created an alarm without rising volume after I saw the previous comment this morning, and it worked correctly (my phone had been asleep for about an hour). The same thing fixes the silent alarms on the stock LineageOS/AOSP clock, so I'm guessing this is an exception that is being handled differently by the two clocks.

**Comment by AhsanSarwar45** [comment]
> I deleted the old alarm that was failing and created a new one without rising volume and alarm worked this morning. Might be luck. Not sure.

If you turn on rising volume, and create an instant alarm (which rings after 5 second, button can be turned on through settings -> developers options -> show instant alarm button), does the alarm still not work?

**Comment by aronovgj** [comment]
Turn on rising volume where? Do I set it in default settings or do I create an alarm first?

I turned it on in the default settings and it works with instant alarm. It just fails after not touching it for a longer time. As already mentioned, after about one hour it still works but after a whole night it doesn't. But there is a notification showing that the alarm rang. It is like the app fails to increase the volume.

**Comment by Rafee-M** [comment]
> t would also be helpful if you guys can suggest opensource apps that work correctly

1. Simple Clock 
2. Fairemail (uses persistent notification)
3. tasks.org (Reminders)

**Comment by jfmcbrayer** [comment]
Instant alarm with rising volume works for me. Overnight with rising volume does not.

**Comment by jfmcbrayer** [comment]
And I'm afraid I spoke too soon. Even with rising volume turned off, my alarms have failed to go off for the last two days.

**Comment by aronovgj** [comment]
After like two weeks it failed again for me

**Comment by TheLastZombie** [comment]
This is a complete shot in the dark, but if you open the Android system settings and go to Apps → Special app access (might be located somewhere else on non-stock Android ROMs) there's a permission called Alarms and reminders, with the following description:

![ResizedImage_Screenshot_20240509-080008(1344)x(579)2024-05-09_08-00-21_4398](https://github.com/vicolo-dev/chrono/assets/8181990/29b22049-f067-4573-90a6-761a628f3040)

Maybe Chrono needs to request this permission? It currently doesn't.

**Comment by theotheroracle** [comment]
also having this issue on lineage android 14, on a oneplus 6, battery optimization disabled

**Comment by Rafee-M** [comment]
> go to Apps → Special app access (might be located somewhere else on non-stock Android ROMs)

Not available on OxygenOS 13

**Comment by AhsanSarwar45** [comment]
> This is a complete shot in the dark, but if you open the Android system settings and go to Apps → Special app access (might be located somewhere else on non-stock Android ROMs) there's a permission called Alarms and reminders, with the following description:
> 
> ![ResizedImage_Screenshot_20240509-080008(1344)x(579)2024-05-09_08-00-21_4398](https://private-user-images.githubusercontent.com/8181990/329142337-29b22049-f067-4573-90a6-761a628f3040.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0NDQ1MzgsIm5iZiI6MTcxNTQ0NDIzOCwicGF0aCI6Ii84MTgxOTkwLzMyOTE0MjMzNy0yOWIyMjA0OS1mMDY3LTQ1NzMtOTBhNi03NjFhNjI4ZjMwNDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTFUMTYxNzE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTVhNDUxOTQ5OWU4ZDZkNzJkMmYwN2NhZGY4YzZhOTk1Y2Y2NTMwZWRkNWU1MjJlMmYyZDMyZGQwMmU0NzNjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qKElUXV3xMvdzS8DifN4iP_qhhEjdq7N24ALOcM8XIs)
> 
> Maybe Chrono needs to request this permission? It currently doesn't.

I think this refers to the `SCHEDULE_EXACT_ALARM` permission, which is only required for apps that aren't alarm clock apps. For alarm clock apps, this permission isn't needed and we instead use `USE_EXACT_ALARM` which is granted automatically by the system. Anyhow, if setting an alarm 1 minute into the future rings correctly, this likely isn't the issue in your case.

**Comment by AhsanSarwar45** [comment]
Just a bit of update, all my current research does lead to this being a battery optimization issue specific to some roms, which would need either a foreground service, or something else that I'm not aware of, to work correctly. I've planned out how the foreground service will be implemented. Currently finishing up the 0.5.0 version which brings widgets and localization and language support. After that, all my focus will be on implementing foreground service and fixing this issue.

**Comment by jgilmore** [comment]
There's an open source clock/alarm app that works at [here](https://github.com/BlackyHawky/Clock)
 Apache 2.0 licensed. Might want to look at what they're doing?

Yes, I'm having these issues. Tried everything that "don't kill my app" recommends. And the test returns 100%. Alarms almost never work. Neither do timers. On a moto stylus 4G 2021 running android 11.

So I'm using "Clock" now.

**Comment by TBat** [comment]
Getting this same problem on Pixel 6, GrapheneOS on Android 14.

Timers won't go off after the screen times out. I can see the notification icon on the lock screen, but it won't ring with the fullscreen alert until I unlock my phone.

**Comment by STVGecko** [comment]
This is a continuation of the #286 

-------

Is your feature request related to a problem? Please describe.
A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]

I don't know why, alarms often don't go off, usually when they're close by.

Describe the solution you'd like
On several other alarm applications, it is possible to define an application as "alarm and reminder" in "special access" (see photo below).

It says:

"Alarms and reminders

Allow this application to set alarms and schedule actions at certain times.
It will then run in the background, which can
battery.

If authorisation is disabled, existing alarms and events
alarms and events programmed by the
the app will not work."

![334299083-be69bfa0-10da-44f8-9554-be29b329cbc2](https://github.com/vicolo-dev/chrono/assets/71868745/e728b1b0-29a6-453f-9bda-d0233bdb6a3f)
-------------
Off-topic: do you know how to reduce the size of photos on github for this type of issues ?

**Comment by joedavid0** [comment]
I have also encountered this issue on my Pixel 8 with GrapheneOS. Already two times this week, the alarm did not go off at all

**Comment by AhsanSarwar45** [comment]
Hi everyone. I have added option to enable foreground notification in [v0.5.2-beta1](https://github.com/vicolo-dev/chrono/releases/tag/v0.5.2-beta1). You can turn it one by going to `Settings > General > Reliability > Show Foreground Notification`. Please try it and let me know if it helps in any ways. 

Also let me know if you experience increased battery consumption by turning on foreground notification (compared to having it off). A little bit increase is expected, but too much increase will be an issue so I will try dialing down the frequency in that case.

I have also added another alarm permission. The docs said it is not needed on newer android versions so I had disabled it on new Android versions previously, but I have re-enabled it now, because why not. It might also help.

**Comment by AhsanSarwar45** [comment]
Not sure if this will fix issue on devices where alarm does not ring after screens time out. Let me know if thats the case.

**Comment by STVGecko** [comment]
> Hi everyone. I have added option to enable foreground notification in [v0.5.2-beta1](https://github.com/vicolo-dev/chrono/releases/tag/v0.5.2-beta1). You can turn it one by going to `Settings > General > Reliability > Show Foreground Notification`. Please try it and let me know if it helps in any ways.
> 
> Also let me know if you experience increased battery consumption by turning on foreground notification (compared to having it off). A little bit increase is expected, but too much increase will be an issue so I will try dialing down the frequency in that case.
> 
> I have also added another alarm permission. The docs said it is not needed on newer android versions so I had disabled it on new Android versions previously, but I have re-enabled it now, because why not. It might also help.

Hello, having tested the application thoroughly, I'm starting to get some answers about the alarms that don't ring.

First of all, you need to know that if you activate the "Rising Volume" option in Settings > Alarm >Default Settings > Sound & Vibration > More > Rising Volume, the alarm sounds progressively, as its name suggests. As a result, there is a risk that the alarm will not be heard.

Then there's the 'Audio Channel' option (same path as above), which I recommend setting to 'alarm' so that you don't unintentionally deactivate it.

Finally, it's good to know that the "Vibration" option (same path as above) works even if you activate the hidden android option "Sensor off". 
https://www.guidingtech.com/turn-off-android-phone-sensors/

I hope these setting tips help. My phone is a Samsung S22.

**Comment by aronovgj** [comment]
Little bit confused about the service notification. In other apps I am used to being able to deactivate the notification while the service still keeps running so that the notification isn't there all the time. I deactivated the notification and the alarm did not go off. But it sounds like the notification itself is supposed to stop the issue instead of a service?

**Comment by AhsanSarwar45** [comment]
> Little bit confused about the service notification. In other apps I am used to being able to deactivate the notification while the service still keeps running so that the notification isn't there all the time. I deactivated the notification and the alarm did not go off. But it sounds like the notification itself is supposed to stop the issue instead of a service?

Can you give examples of apps where this is possible? Because as far as I am aware, Android doesn't allow any foreground service to run without the notification itself because it would defeat the entire purpose of this functionality.

**Comment by aronovgj** [comment]
> > Little bit confused about the service notification. In other apps I am used to being able to deactivate the notification while the service still keeps running so that the notification isn't there all the time. I deactivated the notification and the alarm did not go off. But it sounds like the notification itself is supposed to stop the issue instead of a service?
> 
> Can you give examples of apps where this is possible? Because as far as I am aware, Android doesn't allow any foreground service to run without the notification itself because it would defeat the entire purpose of this functionality.

Yes, one example is K9. It was rewritten some time ago and they had issues with it syncing in background until they introduced the push service. It still works with notification off. Here are their docs regarding this push notification and how to disable it. Below they have a paragraph explaining that you can disable push. So disabling the notification is definitely not the same as disabling the service. 

https://docs.k9mail.app/en/6.400/settings/push/#:~:text=Option%201%3A%20Turn%20off%20the%20notification,-Android%20requires%20apps&text=To%20do%20so%2C%20long%2Dpress,the%20Synchronize%20(Push)%20option.

**Comment by aronovgj** [comment]
Alarm didn't go off with notification turned on either.

**Comment by jfmcbrayer** [comment]
I'm sorry to say that it didn't go off for me, either. I gave it a couple of days to be sure.

**Comment by aronovgj** [comment]
https://github.com/yuriykulikov/AlarmClock

This one works for me too reliably. I think it doesn't even use a foreground service.

**Comment by AhsanSarwar45** [comment]
Unfortunately, its currently not entirely possible for me to see exactly whats causing this, as none of the devices I own are able to reproduce it. If anyone facing the issue is comfortable with doing dev on flutter and android, that would be appreciated.

**Comment by theotheroracle** [comment]
I'm familiar with flutter Dev and have an affected device, how can I help ?

**Comment by AhsanSarwar45** [comment]
> I'm familiar with flutter Dev and have an affected device, how can I help ?

Sorry for the late reply. Its a pretty involved process where you will have to debug what's causing it to not ring on your device. Whether its because of the alarm isolate failing to trigger on time, or not triggering at all, or throwing some other error. You can use `flutter logs` and observe the logs at the time the alarm was supposed to ring. You can also check the source code of other such apps that are working on your device, and try seeing what they are doing differently, difference in permissions, services etc, and then trying to add those permissions to Chrono to see whether it makes a difference.

**Comment by mtalexan** [comment]
I'm having the same problem on a Pixel Fold running CalyxOS.

After the Android 14 upgrade, a lot of my apps started having problems like this, whether I was using Stock ROM, GrapheneOS, or CalyxOS.  

I'm having trouble finding the official reference, but I believe Android 14 introduced a massive change to battery optimization and background apps.
There used to be a radio button for users to pick between three options: off, optimized, and unoptimized. As of Android 14, I believe they say "unoptimized" is now no longer an option, there's only a toggle to pick between the former "off" and "optimized".  
Associated with this change, they added some exceptions though. Certain behaviors can cause the "optimized" state be treated like "unoptimized" for that app, but I don't remember exactly what was included. The one key behavior they did mention was foreground notifications, but they changed how even those qualified. I believe minimum priority persistent notifications no longer qualify. 

So if I'm not mistaken, I believe Android 14 rules are now that you need a foreground service with a persistent notification at non-minimum priority, and the user is free to modify that notification to set it to be silent and configure all their persistent silent notifications to be minimized if they don't want it visible in the status bar at all times.

**Comment by STVGecko** [comment]
I think I've found the problem!

My alarm didn't go off either, specifically around 7, 8 or even 9 o'clock. In my case, I had an alarm set for 8.14am that wasn't ringing.

In fact I've realised that my alarms only go off if their "type" is set to "Daily" and not "On Specified week days".

As you'll see from the attached photos, I've set the alarm to 7.34am so that it sounds on Mondays, Tuesdays, Wednesdays, Thursdays and Fridays. However, the alarm doesn't sound on these days and displays a notification which, as mentioned in a previous comment.

@Rafee-M :
>  A notification for upcoming alarm at 11:15 AM (set at 10:30 AM) was receieved. But the alarm failed to ring. Opening the app also didn't make the alarm ring as well.

Here's how I solved the problem: 

-click on the problematic alarm/ringtone
-Schedule>Type>Click on the arrow
-Select "Daily"

![1](https://github.com/user-attachments/assets/9934356a-b5c7-493d-ac1a-29e209f0a855)

![2](https://github.com/user-attachments/assets/19bca742-94c2-4f5d-9bc9-54735a40ae39)

![3](https://github.com/user-attachments/assets/e51111dc-d6f7-44cd-b136-0ee095b2ba95)

-> as a result, the alarm sounds (I've managed to sound an alarm at 7.34am, 8.24am and 10am.

When in doubt, I also recommend activating the alarm's "vibration" function.

It's also good to know that the "on specific dates" type works.

Don't hesitate to try this method on your own, in the hope that it will help you. Have a nice day :)

**Comment by mtalexan** [comment]
I wouldn't call that a solution (use "Daily instead of Specific Days"), but it does narrow down the problem.

**Comment by aronovgj** [comment]
I can confirm that it seems to work OK for daily but not for specific days.

**Comment by aronovgj** [comment]
I have been testing v0.6.0-beta1 for two weeks and alarms for specific days now also has been working reliantly, with foreground notification turned *off*.

Can anybody confirm that the issue is resolved without having to turn on the foreground notification?

**Comment by mtalexan** [comment]
Can confirm it seems to be working for me now too. I'm also on v0.6.0-beta1 but I have the foreground service enabled and set to silent+minimized. The alarms set for only certain days of the week are working now.

**Comment by aronovgj** [comment]
> Can confirm it seems to be working for me now too. I'm also on v0.6.0-beta1 but I have the foreground service enabled and set to silent+minimized. The alarms set for only certain days of the week are working now.

Please also check if it works now with the service off

**Comment by mtalexan** [comment]
Confirmed. It works consistently with the foreground service disabled too.

**Comment by joedavid0** [comment]
I have now also tested the Beta for two weeks and the good news is that the alarm is now reliable and thus useable.

I have however noticed that the alarm aswell as the _timer_ still almost always have delays of 20-30s, in rare cases up to a minute. In case of a daily alarm, this is useable, in case of a short-term timer it is not however :/

**Comment by mtalexan** [comment]
I haven't had an issue like that, my alarms are triggering right on time

**Comment by Rafee-M** [comment]
I'm happy to say they are working for me as well! Timers do seem to delayed on and off. I noticed alarms fail once with an error code on one occasion. I'll try to reproduce it

**Comment by AhsanSarwar45** [comment]
Beta1 introduced a background service, which might have been responsible for alarms now being reliable. On the other hand, it might also be responsible for [this](https://github.com/vicolo-dev/chrono/issues/355) 'error decoding' bug. In [beta2](https://github.com/vicolo-dev/chrono/releases/tag/v0.6.0-beta2), I have disabled the background service by default. I do think that fixed the error, but let me know if alarms continue ringing reliably without that background service or not.

**Comment by aronovgj** [comment]
Alarms continue ringing reliably with background service being off

**Comment by jackusay** [comment]
Does your beta2 show beta1 in about page?

**Comment by aronovgj** [comment]
Yes it does

**Comment by Rafee-M** [comment]
Still the issue seems to persist. With the Foreground or Background service, alarms seems to not ring.

Interestingly, upcoming alarm notifications (5 minutes before) seems to work fine, so I assume the app is awake? I even tried with keeping the app in the recents page

**Comment by PitiBouchon** [comment]
might be the same issue as https://github.com/vicolo-dev/chrono/issues/236

**Comment by codychen98** [comment]
Might be a dumb question, but what if you give this app the ability to be set as "Device Admin App"? I thought that will make it always active, so for folks with this issue, they can optionally grant this access?
